### PR TITLE
BFD-3589: CI Improvements for Shell Scripts (ShellCheck)

### PR DESCRIPTION
**Pull request by dondevun**
**JIRA Ticket:**
BFD-3589

### What Does This PR Do?

Adds pre-commit hook to run shellcheck against newly committed shell scripts, and a GitHub Action to run shellcheck against shell scripts in a PR.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

### What Security Implications Does This PR Have?

Please indicate if this PR does any of the following:  

* Adds any new software dependencies
* Modifies any security controls
* Adds new transmission or storage of data
* Any other changes that could possibly affect security? 

* [x] I have considered the above security implications as it relates to this PR. (If one or more of the above apply, it cannot be merged without the ISSO or team security engineer's (`@sb-benohe`) approval.) 

### Validation

Have you fully verified and tested these changes? Is the acceptance criteria met? Please provide reproducible testing instructions, code snippets, or screenshots as applicable.
   
Test by attempting to commit an erroneous shell script.

### CODE REVIEW

**Review by sb-benohe — COMMENTED** [review]
LGTM: Like the use of `filename##*.`  and the check for Hashbangs that should cover most situations for the use of bash.

File: .github/scripts/pre-commit.sh [hunk 1]
```diff
@@ -49,4 +49,31 @@ checkSecretFilesForPlainText() {
   set -e
 }
 
+# Runs shellcheck against any committed shell scripts
+runShellCheckForCommitFiles() {
+  echo 'Running shellcheck on committed files'
+  if [ "$(which shellcheck)" ]; then
+    tmpfile=$(mktemp)
+    git diff --cached --name-only --diff-filter=ACM >"$tmpfile"
+    commits=`cat $tmpfile`
```
Comment (malessi) [inline .github/scripts/pre-commit.sh:58]:
```suggestion
    commits=$(cat "$tmpfile")
```

File: .github/scripts/pre-commit.sh [hunk 2]
```diff
@@ -49,4 +49,31 @@ checkSecretFilesForPlainText() {
...
+    for file in $commits; do
+      firstTwo=$( sed 's/^\(..\).*/\1/;q' "$file" )
+      filename=$(basename -- "$file")
+      extension="${filename##*.}"
+      # check for a hashbang or a .sh extension to determine if this is a shell script.
+      if [ "$firstTwo" == "#!" ] || [ $extension == "sh" ]; then
```
Comment (malessi) [inline .github/scripts/pre-commit.sh:64]:
```suggestion
      if [ "$firstTwo" == "#!" ] || [ "$extension" == "sh" ]; then
```

File: .github/scripts/pre-commit.sh [hunk 3]
```diff
@@ -49,4 +49,31 @@ checkSecretFilesForPlainText() {
...
+        # run shellcheck with severity level warning, and suppress warnings about invalid hashbangs (allows it to ignore other types of scripts, e.g. python)
+        shellcheck -e SC1071,SC2239 -S warning "$file"
+        if [[ $? != 0 ]]; then
```
Comment (malessi) [inline .github/scripts/pre-commit.sh:67]:
```suggestion       
        if ! shellcheck -e SC1071,SC2239 -S warning "$file"; then
```

**Review by malessi — COMMENTED** [review]
Just a couple of stylistic/bash best practice things that really aren't a big deal. This looks great!

**Comment by aschey-forpeople** [comment]
Mind filling out the validation section on the PR template? Thanks!

**Review by sb-benohe — APPROVED** [review]

**Review by aschey-forpeople — APPROVED** [review]

**Pull request merged**